File: src/python_parameter_event_handler/python_parameter_event_handler/parameter_event_handler.py
import rclpy
from rclpy.node import Node
import rclpy.parameter
from rcl_interfaces.msg import Parameter, ParameterEvent

from rclpy.parameter_event_handler import ParameterEventHandler

class SampleNodeWithParameters(Node):
    def __init__(self):
        super().__init__("node_with_parameters")

        self.declare_parameter("an_int_param", 0)
        self.declare_parameter("another_double_param", 0.0)

        self.handler = ParameterEventHandler(self)

        self.callback_handle = self.handler.add_parameter_callback(
            parameter_name="an_int_param",
            node_name=self.get_name(),
            callback=self.callback,
        )

        self.handler.add_parameter_callback(
            "a_double_param",
            "parameter_blackboard",
            self.callback,
        )

        self.event_callback_handle = self.handler.add_parameter_event_callback(
            self.event_callback
        )

    def callback(self, p: Parameter) -> None:
        self.get_logger().info(
            "Received an update to parameter: {}: {}".format(
                p.name, rclpy.parameter.parameter_value_to_python(p.value)
            )
        )

    def event_callback(self, parameter_event: ParameterEvent):
        self.get_logger().info(f"Type: {type(parameter_event)}")
        self.get_logger().info(f"Type: {type(parameter_event.changed_parameters)}")
        self.get_logger().info(f"Type: {type(parameter_event.changed_parameters[0])}")

        self.get_logger().info(f"Received parameter event from node {parameter_event.node}")

        p: Parameter
        for p in parameter_event.changed_parameters:
            self.get_logger().info(
                "Inside event: {} changed to: {}".format(
                    p.name, rclpy.parameter.parameter_value_to_python(p.value)
                )
            )

def main():
    rclpy.init()
    node = SampleNodeWithParameters()
    rclpy.spin(node)
    rclpy.shutdown()

File: src/cpp_parameter_event_handler/src/parameter_event_handler.cpp
#include <memory>

#include "rclcpp/rclcpp.hpp"

class SampleNodeWithParameters : public rclcpp::Node {
public:
    SampleNodeWithParameters() : Node("node_with_parameters") {
        this->declare_parameter<int>("an_int_param", 0);
        this->declare_parameter<double>("another_double_param", 0.0);

        // Create a parameter subscriber that can be used to monitor parameter changes
        // (for this node's parameters as well as other nodes' parameters)
        param_subscriber_ = std::make_shared<rclcpp::ParameterEventHandler>(this);

        // Set a callback for this node's interget parameter, "an_int_param"
        auto cb_lambda = [this] (const rclcpp::Parameter &p) { cb(p); };
        cb_handle_ = param_subscriber_->add_parameter_callback("an_int_param", cb_lambda);

        // Now, add a callback to monitor any changes to the remote node's parameter.
        // In this case, we supply the remote node name.
        auto cb2_lambda = [this] (const rclcpp::Parameter &p) { cb2(p); };
        auto remote_node_name = std::string("parameter_blackboard");
        auto remote_param_name = std::string("a_double_param");
        cb2_handle_ = param_subscriber_->add_parameter_callback(
            remote_param_name, cb2_lambda, remote_node_name
        );

        auto event_cb_lambda =
            [this] (const rcl_interfaces::msg::ParameterEvent &param_event) { event_cb(param_event); };
        event_cb_handle_ = param_subscriber_->add_parameter_event_callback(event_cb_lambda);
    }

private:
    std::shared_ptr<rclcpp::ParameterEventHandler> param_subscriber_;
    std::shared_ptr<rclcpp::ParameterCallbackHandle> cb_handle_;
    std::shared_ptr<rclcpp::ParameterCallbackHandle> cb2_handle_;
    std::shared_ptr<rclcpp::ParameterEventCallbackHandle> event_cb_handle_;

    void cb(const rclcpp::Parameter &p) {
        RCLCPP_INFO(this->get_logger(),
            "cb: Received an update to parameter \"%s\" of type %s: \"%ld\"",
            p.get_name().c_str(),
            p.get_type_name().c_str(),
            p.as_int()
        );
    }

    void cb2(const rclcpp::Parameter &p) {
        RCLCPP_INFO(this->get_logger(),
            "cb2: Received an update to parameter \"%s\" of type %s: \"%.02lf\"",
            p.get_name().c_str(),
            p.get_type_name().c_str(),
            p.as_double()
        );
    }

    void event_cb(const rcl_interfaces::msg::ParameterEvent &param_event) {
        RCLCPP_INFO(this->get_logger(),
            "Received parameter event from node \"%s\"",
            param_event.node.c_str()
        );
        for (const auto &p : param_event.changed_parameters) {
            RCLCPP_INFO(this->get_logger(),
                "Inside event: \"%s\" changed to %s",
                p.name.c_str(),
                rclcpp::Parameter::from_parameter_msg(p).value_to_string().c_str()
            );
        }
    }
};

int main(int argc, char **argv) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<SampleNodeWithParameters>());
    rclcpp::shutdown();
    return 0;
}